qiskit_optimization/problems/quadratic_objective.py

```diff
@@ -162,7 +163,17 @@ def evaluate(self, x: Union[ndarray, List, Dict[Union[int, str], float]]) -> flo
 
         Returns:
             The value of the quadratic objective given the variable values.
+
+        Raises:
+            QiskitOptimizationError: if the shape of the objective function does not match with
+                the number of variables.
         """
+        n = self.quadratic_program.get_num_vars()
+        if self.linear.coefficients.shape != (1, n) or self.quadratic.coefficients.shape != (n, n):
+            raise QiskitOptimizationError(
+                "The shape of the objective function does not match with the number of variables. "
+                "Need to define the objective function after defining all variables"
+            )
```

Comment:
Just curious, what if I have a variable that is only in a constraint, but not in objective. Would it work? I guess it should.

Reply:
Yes, it works fine with this PR too.
```python
from qiskit_optimization import QuadraticProgram

qp = QuadraticProgram()
qp.binary_var_list(3)
qp.linear_constraint({0: 1, 1: 1, 2: 1}, '==', 1)
print(qp)
print(qp.objective.evaluate([0, 0, 0]))
```
```
# output
\ This file has been generated by DOcplex
\ ENCODING=ISO-8859-1
\Problem name: CPLEX

Minimize
 obj:
Subject To
 c0: x0 + x1 + x2 = 1

Bounds
 0 <= x0 <= 1
 0 <= x1 <= 1
 0 <= x2 <= 1

Binaries
 x0 x1 x2
End

Traceback (most recent call last):
  File "obj2.py", line 9, in <module>
    print(qp.objective.evaluate([0, 0, 0]))
  File "/Users/ima/tasks/1_2021/qiskit/opt/qiskit_optimization/problems/quadratic_objective.py", line 173, in evaluate
    raise QiskitOptimizationError(
qiskit_optimization.exceptions.QiskitOptimizationError: 'The shape of the objective function does not match with the number of variables. Need to define the objective function after defining all variables'
```

Reply:
I mean slightly different. Suppose we have an objective, but not all variables are present in the objective. Some of them are only in the constraints for this problem.

Reply:
Do you mean this?
```python
from qiskit_optimization import QuadraticProgram

qp = QuadraticProgram()
qp.binary_var_list(3)
qp.minimize(linear={0: 1})
qp.linear_constraint({0: 1, 1: 1, 2: 1}, '==', 1)
print(qp)
print(qp.objective.evaluate([0, 0, 0]))
```
```
# output
\ This file has been generated by DOcplex
\ ENCODING=ISO-8859-1
\Problem name: CPLEX

Minimize
 obj: x0
Subject To
 c0: x0 + x1 + x2 = 1

Bounds
 0 <= x0 <= 1
 0 <= x1 <= 1
 0 <= x2 <= 1

Binaries
 x0 x1 x2
End

0.0
```

Reply:
Yes, thanks a lot!